### ISSUE TITLE
cmd/compile: pre-allocate a slice of provable length 

### ISSUE THREAD

**Issue by dsnet** [golang/go#64595]
### Go version

go1.21.2

### What operating system and processor architecture are you using (`go env`)?

```shell
n/a
```

### What did you do?

Compile the following snippet:
```go
func transform(a []int) []uint {
	var b []uint
	for _, x := range a {
		b = append(b, uint(x))
	}
	return b
}
```

### What did you expect to see?

I expect to see the following:
* An implicit call to `makeslice` of a given length and capacity. The elements within the length need not be explicitly zeroed (since it can be provably overwritten).
*  The `append` call in the loop be essentially rewritten as an index operation, thus avoiding a call to `runtime.growslice` in every iteration of the loop.

The conversation of `int` to `uint` is for demonstration, but the above optimization should be doable for any conversion that does not panic.

### What did you see instead?

Right now, this compiles more or less to what you see in the code.

However, I would expect it to compile to something like:
```go
func transform(a []int) []uint {
	b := make([]uint, len(a), len(a) + ...) // where ... is some extra capacity to reproduce above behavior
	for i, x := range a {
		b[i] = uint(x)
	}
	return b
}
```

**Comment by cagedmantis** [comment]
cc [@golang/compiler](https://github.com/orgs/golang/teams/compiler)

**Comment by Jorropo** [comment]
There is also a linter which helps with this https://github.com/alexkohler/prealloc, it's not good enough to do machine fixes, but it's way good enough for primates look at it's output and fix interesting cases.
Primates easily have the missing context of which panic, error and other branches which are worth disregarding and worth optimizing.

**Comment by randall77** [comment]
I believe that use case is covered by upcoming iterator proposals, in particular `iter.Map` from #61898 and `slices.Append` from #61899.

**Comment by AlexanderYastrebov** [comment]
@randall77 Thank you. Those revolve around iter.Seq which does not provide size so they will not be able to pre-allocate result afaict.

**Comment by randall77** [comment]
True. The size could be passed somehow using an unexported method, although I don't immediately see an obvious way to do that (we could return a named `iter.Seq` with its own methods, but because `iter.Seq` must be a function type, maybe there's no place to store the length to return? Except hackily in the closure somehow).

**Comment by dsnet** [comment]
> There is also a linter which helps

The issue at hand is that most times I don't want to think about it when writing code.
On the coding side of this, I want to focus on the business logic, rather than optimizations.
On the optimization side, writing the loop naturally with append allows PGO to make a decision whether preallocating is worth it if it realizes the loop is continually re-allocating in practice.

**Comment by prattmic** [comment]
Several of us have recently been brainstorming online slice/map pre-sizing (#64703). To a degree, this is a special case of that, where the ideal [1] size can be determined statically.

[1] The exception is that `transform` returns `b`. If the caller continues to append to `b`, then the ideal initial capacity is even higher. Online tracking could capture that.

**Comment by adonovan** [comment]
I wonder how often this optimizations makes things slower: If `a` is a very large slice of empty slices, then the new algorithm makes two passes over it, causing twice as many cache misses. Rare, I suspect.
